pkg/tanka/inline.go

```diff
@@ -0,0 +1,92 @@
+package tanka
+
+import (
+	"encoding/json"
+	"fmt"
+	"path/filepath"
+
+	"github.com/grafana/tanka/pkg/jsonnet/jpath"
+	"github.com/grafana/tanka/pkg/kubernetes/manifest"
+	"github.com/grafana/tanka/pkg/process"
+	"github.com/grafana/tanka/pkg/spec"
+	"github.com/grafana/tanka/pkg/spec/v1alpha1"
+)
+
+// InlineLoader loads an environment that is specified inline from within
+// Jsonnet. The Jsonnet output is expected to hold a tanka.dev/Environment type,
+// Kubernetes resources are expected at the `data` key of this very type
+type InlineLoader struct{}
+
+func (i *InlineLoader) Load(path string, opts JsonnetOpts) (*v1alpha1.Environment, error) {
+	raw, err := EvalJsonnet(path, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	var data interface{}
+	if err := json.Unmarshal([]byte(raw), &data); err != nil {
+		return nil, err
+	}
+
+	envs, err := extractEnvs(data)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(envs) > 1 {
+		names := make([]string, 0, len(envs))
+		for _, e := range envs {
+			names = append(names, e.Metadata().Name())
+		}
+		return nil, ErrMultipleEnvs{path, names}
+	}
+
+	if len(envs) == 0 {
+		return nil, fmt.Errorf("Found no environments in '%s'", path)
+	}
+
+	root, base, err := jpath.Dirs(path)
+	if err != nil {
+		return nil, err
+	}
+
+	name, err := filepath.Rel(root, base)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Re-serializing the entire env here. This is horribly inefficient
+	envData, err := json.Marshal(envs[0])
+	if err != nil {
+		return nil, err
+	}
+
+	env, err := spec.Parse(envData, name)
```

Comment:
We could change `spec.Parse` to expect a `manifest.Manifest` object?

Reply:
I fear that is actually not as easy as it sounds.

The whole reason we use the `json` re-serialization is to do a deep copy of the underlying data, which is highly tricky in Go otherwise due to pointer types, etc. The resulting `v1alpha1.Environment` must not be affected by future changes to the original `manifest.Manifest`.

I'd say we first benchmark how bad this actually is, it might be easily outweighed by an order of magnitude by the Jsonnet compiler, so this might not matter at all